File: client/app/src/main/java/com/androidmac/client/protocol/Messages.kt
package com.androidmac.client.protocol

import org.json.JSONArray
import org.json.JSONObject

data class ScreenInfo(val width: Int, val height: Int, val dpi: Int)

data class ClientHello(
    val device: String,
    val screen: ScreenInfo,
    val capabilities: List<String>,
    val codecs: List<String>,
    val connectionType: String = "wifi"
) {
    fun toJson(): String {
        val obj = JSONObject()
        obj.put("device", device)
        obj.put("screen", JSONObject().apply {
            put("width", screen.width)
            put("height", screen.height)
            put("dpi", screen.dpi)
        })
        obj.put("capabilities", JSONArray(capabilities))
        obj.put("codecs", JSONArray(codecs))
        if (connectionType != "wifi") {
            obj.put("connectionType", connectionType)
        }
        return obj.toString()
    }
}

data class DisplayInfo(val width: Int, val height: Int)

data class ServerHello(
    val virtualDisplay: DisplayInfo,
    val codec: String,
    val bitrate: Int,
    val fps: Int,
    val streamPort: Int,
    val touchPort: Int,
    val videoPort: Int = 0
) {
    companion object {
        fun fromJson(json: String): ServerHello {
            val obj = JSONObject(json)
            val disp = obj.getJSONObject("virtualDisplay")
            return ServerHello(
                virtualDisplay = DisplayInfo(disp.getInt("width"), disp.getInt("height")),
                codec = obj.getString("codec"),
                bitrate = obj.getInt("bitrate"),
                fps = obj.getInt("fps"),
                streamPort = obj.getInt("streamPort"),
                touchPort = obj.optInt("touchPort", 0),
                videoPort = obj.optInt("videoPort", 0)
            )
        }
    }
}

data class ClientReady(val udpPort: Int) {
    fun toJson(): String {
        val obj = JSONObject()
        obj.put("udpPort", udpPort)
        return obj.toString()
    }
}

File: server/internal/protocol/messages.go
package protocol

// ClientHello is sent by Android to Mac during handshake
type ClientHello struct {
	Device         string     `json:"device"`
	Screen         ScreenInfo `json:"screen"`
	Capabilities   []string   `json:"capabilities"`
	Codecs         []string   `json:"codecs"`
	ConnectionType string     `json:"connectionType,omitempty"` // "wifi" (default) or "usb"
	Bitrate        int        `json:"bitrate,omitempty"`        // requested bitrate in bps, 0 = server default
}

// IsUSB returns true if the client is connected via USB/ADB.
func (h *ClientHello) IsUSB() bool {
	return h.ConnectionType == "usb"
}

type ScreenInfo struct {
	Width  int `json:"width"`
	Height int `json:"height"`
	DPI    int `json:"dpi"`
}

// ServerHello is sent by Mac to Android in response
type ServerHello struct {
	VirtualDisplay DisplayInfo `json:"virtualDisplay"`
	Codec          string      `json:"codec"`
	Bitrate        int         `json:"bitrate"`
	FPS            int         `json:"fps"`
	StreamPort     int         `json:"streamPort"`
	TouchPort      int         `json:"touchPort"`
	VideoPort      int         `json:"videoPort,omitempty"` // TCP video port for USB mode
}

type DisplayInfo struct {
	Width  int `json:"width"`
	Height int `json:"height"`
}

// ClientReady is sent by Android to Mac after binding the UDP socket.
// For USB mode, UDPPort is 0 (video uses TCP instead).
type ClientReady struct {
	UDPPort int `json:"udpPort"`
}

// Heartbeat is sent periodically on the control channel
type Heartbeat struct {
	Timestamp int64 `json:"timestamp"`
}
